File: 369__Plus_One_Linked_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def plusOne(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        def addone(root):
            if not root:
                return 1
            add_on_prev_digit = addone(root.next)
            root.val += add_on_prev_digit
            add_on_prev_digit = root.val / 10
            root.val %= 10
            return add_on_prev_digit

        if not head:
            return head

        all_9 = addone(head)

        if all_9 > 0:
            root = ListNode(1)
            root.next = head
        else:
            root = head
        return root

File: Java/369__Plus_One_Linked_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode plusOne(ListNode head) {
        if (head == null) {
            return null;
        }
        head = reverse(head);
        int incre = 1, old_val;
        ListNode ptr = head;
        while (ptr != null) {
            old_val = ptr.val;
            ptr.val = (old_val + incre) % 10;
            incre = (old_val + incre) / 10;
            ptr = ptr.next;
        }
        head = reverse(head);
        if (incre == 1) {
            ListNode result = new ListNode(1);
            result.next = head;
            return result;
        } else {
            return head;
        }
    }

    private ListNode reverse(ListNode head) {
        ListNode rest = head, temp = null, result = null;
        while (rest != null) {
            temp = rest.next;
            rest.next = result;
            result = rest;
            rest = temp;
        }
        return result;
    }
}
